pkg/storegateway/bucket.go

```diff
@@ -862,7 +860,7 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 		}
 	}
 
-	gspan, gctx := tracing.StartSpan(gctx, "bucket_store_preload_all")
+	span, ctx := tracing.StartSpan(ctx, "bucket_store_preload_all")
 
 	s.mtx.RLock()
```

Comment:
Do we need to take this lock to run `blocks := s.blockSet.getFor(req.MinTime, req.MaxTime, req.MaxResolutionWindow, reqBlockMatchers)`? `s.blockSet` has its own lock 🤔

Reply:
i think this lock prevents a race in unloading blocks and starting to query them. Unloading blocks is here

https://github.com/grafana/mimir/blob/ec5e7323106f3005124597377ca7a3f6886c9056/pkg/storegateway/bucket.go#L446-L452

I will see what we can do. But I believe this lock is necessary

Reply:
Right. Then we could move the call to `s.blockSet.getFor()` to `synchronousSeriesSet()`?

Reply:
I opened https://github.com/grafana/mimir/pull/3546 with the locking simplification https://github.com/grafana/mimir/pull/3546